### ISSUE TITLE
Dropping a persisted table does not free the persisted data

### ISSUE THREAD

**Issue by ruchirK** [MaterializeInc/materialize#8185]
When you call `DROP TABLE` on a table created with materialize run with `--experimental --persistent-user-tables` the table is removed from the catalog but the underlying data lives on in the `mzdata/persist` directory.

To repro:

run materialize with `cargo run -- -w 4 --log-file=stderr --dev --persistent-user-tables --experimental`

(note that I purposefully omitted --persistent-system-tables` for id clarity)

And now you can do the following sequence of commands

```
materialize=> create table test (f1 text);
CREATE TABLE
materialize=> select * from test;
 f1
----
(0 rows)
```

let's take a moment to laugh at my lack of sql-fu
```
materialize=> insert into test values ("hi there");
ERROR:  column "hi there" does not exist
```
and then go back to insert some data
```
materialize=> insert into test values ('hi there');
INSERT 0 1
materialize=> select * from test;
    f1
----------
 hi there
(1 row)
```

At this point if we inspect the persistent state we see

```
altaria-2:materialize Test$ ls -lrt mzdata/persist/blob/
total 32
-rw-r--r--  1 Test  staff  158 Sep  6 18:57 LOCK
-rw-r--r--  1 Test  staff  120 Sep  6 18:58 Id(0)-trace-6
-rw-r--r--  1 Test  staff  203 Sep  6 18:58 Id(0)-trace-7
-rw-r--r--  1 Test  staff  724 Sep  6 18:58 META
```

`Id(0)` represents the table `test`.  I'm not sure if this information is derivable from some other system view. 

Anyway, now if we delete the table, with

```
DROP TABLE test;
```

the relevant `Id(0)` files stick around forever, even after we add a new table / restart. We have the machinery to evict unnneeded batches, so this should be pretty easy to stitch together.

cc: @danhhz

**Comment by danhhz** [comment]
Good question! @mjibson is our expert on all things related our SQL consistency, so I'll let him chime in.

**Comment by mkysel** [comment]
My observation might not be related to the actual removal of files on disk, so I am sorry if I am mixing things together!

**Comment by mkysel** [comment]
The isolation here has two violations: non-repeatable reads and dirty reads. Depends on what isolation level materialize wants to be. Neither repeatable read nor serializable allow for these two violations.

**Comment by mjibson** [comment]
(We're going to discuss this elsewhere.)

**Comment by danhhz** [comment]
Thoughts in no particular order:
- The relevant persist call once coord is ready to delete the table data is https://github.com/MaterializeInc/materialize/blob/909b3edb6021df85c55276844f7d85119da75387/src/persist/src/indexed/runtime.rs#L364-L373
- It might be odd that this call is on `RuntimeClient` which is a handle to the whole persist system as opposed to being on `StreamWriteHandle`, which a handle to a single persisted collection. We've never been quite sure which one of the two it belonged on, so if it turns out to become clear from hooking things up that in the relevant place you have a `StreamWriteHandle` and don't have a `RuntimeClient`+ the string name identifier, then we can move it.
- I'm not closely familiar with the coord side of this, so you'll probably have to talk to matt (or the sql-coord slack channel) for questions about that
- It's not clear to me whether the DROP TABLE call should block on the full deletion of the data or do it in the background. If the latter, what's our contract on what happens if the background deletion is interrupted by a restart? We could develop a bunch of infra here for keeping track of it durably to ensure it gets kicked off again after a restart, but maybe that's not terribly worthwhile and we should just start with the easy thing of blocking.
- Though maybe that's not so easy after all. One thing to consider there is what happens if the data deletions fail (e.g. storage is down or partially down). We don't want to delete part of the data but keep the SQL definition around (this would be bad if we restarted and reallowed usage of the table with partial data). But if we delete the catalog definition first, then we might end up in the same orphaned data situation again.
- Persist already internally keeps track of the collections it has deleted because reusing the same name could cause some tricky distributed systems problems down the road. Maybe we can solve all of the above by keeping track of which bits of data need to be deleted alongside and persist could itself pick the work back up again after a restart.
- There's certainly complexity I'm missing here but that's what I came up with at a first glance

**Comment by cuongdo** [comment]
There's been no activity on this for a while. Is this still being worked on or on hold while we research answers to the outstanding questions?

**Comment by mkysel** [comment]
@mjibson and I are working on designing a prerequisite system. Doing this right requires a non-trivial amount of work that needs to be properly thought out.

**Comment by mkysel** [comment]
we had some more discussion on this and decided that this work will be split into two deliverables.
1) this issue. We will implement a quick solution that will unblock Milestone 4. The deletion will happen in a non-atomic operation in this order: first, drop the catalog item; second, drop the data on disk. If the materialize process crashes after step one, but before step two, the data on disk will be leaked.
2) we will handle the atomicity issue in a separate SQL task. I will update this comment once I have filed the follow-up.

It has also been agreed that we all prefer to block until the table is fully dropped. This behavior might have edge cases, but in the happy case, the `drop table` statement will block until the operation is successful.

**Comment by danhhz** [comment]
Confirming this is still an issue, but one we'll fix by EA/GA

**Comment by aljoscha** [comment]
@pH14 @sploiselle With the recent changes, it feels like this issue is now actually resolved?

**Comment by danhhz** [comment]
Yeah, I've been intending to hand this over to QA, but I was planning to first do a quick sanity check that it works end-to-end so we don't waste their time if there's some obvious edge case

**Comment by aljoscha** [comment]
sounds good!

**Comment by benesch** [comment]
Do we have a plan for cleaning up already-leaked shards?

**Comment by danhhz** [comment]
@benesch Great Q! Filed that as a separate issue https://github.com/MaterializeInc/materialize/issues/19155

**Comment by danhhz** [comment]
@sploiselle I tried to take this for a spin locally with a bunch of additional logging added by creating and dropping a table, and I don't see the finalization happening. haven't looked at the upper yet, but I don't think I see the since advancing to empty. any advice?

**Comment by benesch** [comment]
> That being said, to mark this as done, we might want to finally dig in and make `StorageResponse::DroppedIds` occur reliably.

I endorse this! Surely there is a small memory leak or something due to the current behavior (in addition to the shard leak).

I've paged out the purpose of `DroppedIds`. Why does the controller need to wait for the cluster to confirm that it has dropped the source/sink in question before taking cleanup action? It seems like the kind of concept that doesn't make sense in a world where source/sink clusters can be replicated. (Do you wait for all replicas to drop the source? What if one of the replicas is perma-busted? Etc.)

**Comment by benesch** [comment]
@ParkMyCar bumped into some issues around this recently. https://github.com/MaterializeInc/materialize/pull/25173 will enable dropping tables' shards. @sploiselle reports that we still leak state for subsources (including progress shards?), but that's something that he's planning to tackle as part of https://github.com/MaterializeInc/materialize/issues/24773.

**Comment by sploiselle** [comment]
https://github.com/MaterializeInc/materialize/pull/26826 closed the technical challenge in making this happen; leaving this issue open because we should add a testdrive test to verify that this occurs https://github.com/MaterializeInc/materialize/blob/909596a50c462d13c43070d0eb66809f7dff1927/src/testdrive/src/action/consistency.rs#L75-L76
